bpo-29463: Add docstring field to some AST nodes.

methane opened the pull request:
ClassDef, ModuleDef, FunctionDef, and AsyncFunctionDef has docstring
field for now.  It was first statement of there body.

haypo commented:
I know that CPython doesn't use GitHub yet, but I created a pull request to ease the review of this large change.

Related CPython issue: http://bugs.python.org/issue17870

eliben commented:
OOC, how does it make it easier to review? Github's code review is pretty bare-bones compared to the Rietvield instance CPython uses (IMHO)

haypo commented:
2016-11-25 14:45 GMT+01:00 Eli Bendersky <notifications@github.com>:
> OOC, how does it make it easier to review? Github's code review is pretty bare-bones compared to the Rietvield instance CPython uses (IMHO)

You can review commits individually.

brettcannon commented:
Closing this to make sure I didn't miss notifying anyone about the migration.

methane committed:
bpo-29463: Add docstring field to some AST nodes.

ClassDef, ModuleDef, FunctionDef, and AsyncFunctionDef has docstring
field for now.  It was first statement of there body.

vstinner commented on Doc/library/ast.rst:
```diff
@@ -133,9 +133,17 @@ and classes for traversing abstract syntax trees:
+
+   .. versionchanged:: 3.7
+      The docstring is exported from attribute of the node, instead of first
+      statement in it's body.
```
*of its* body

methane committed:
fix document.  thanks travis!

vstinner commented on Doc/library/ast.rst:
```diff
@@ -133,9 +133,17 @@ and classes for traversing abstract syntax trees:
+
+   .. versionchanged:: 3.7
+      The docstring is exported from attribute of the node, instead of first
+      statement of its body.
```
Sorry, the doc still sounds strange to me. I propose:

The docstring is now exported from the node ``docstring`` field, instead of the first body statement.

methane replied:
thanks.  In above, I feel "AsyncFunctionDef is added" wrong too.  How do you think?

* ``Added :class:`AsyncFunctionDef` support``
* ``:class:`AsyncFunctionDef` is supported now``

vstinner replied:
Ah, you're right:

:class:`AsyncFunctionDef` is now supported.

vstinner reviewed (approved):
Just a minor nit, not a blocker one.

methane committed:
doc fixes

vstinner commented:
Thanks for the last documentation fix ;-)

Nice enhancement, it should help to make AST code simpler!

Carreau commented:
Thanks for this ! Improvement to the AST are welcome !

Would it have been possible to make the docstring optional ? (It's already breaking things, like IPython). 

Should I comment on upstream bpo ?

methane commented:
@Carreau Yes, please discuss on bpo.  Merged pull request is not good place to discussion,
because it's too hard to search.

vstinner commented:
> Would it have been possible to make the docstring optional ? (It's already breaking things, like IPython).

I created http://bugs.python.org/issue29622

Carreau commented:
Thanks @haypo, responded and subscribed to both.
